src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -235,6 +266,66 @@ internal static void SetProperty(object resource, string propertyName, object va
             }
         }
 
+        internal static void SetInstanceAnnotations(object resource, string annotationPropertyName, ODataDeserializerProvider deserializerProvider,
+            ODataDeserializerContext readContext, ODataResource oDataResource, IDictionary<string, IDictionary<string, object>> instAnnotationDictionary)
+        {
+            bool shouldSetInstanceAnnotations = false;
+            if(oDataResource.InstanceAnnotations != null && oDataResource.InstanceAnnotations.Count > 0)
+            {
+                string annotationKey = string.Empty;
+
+                if (!instAnnotationDictionary.ContainsKey(annotationKey))
+                {
+                    instAnnotationDictionary.Add(annotationKey, new Dictionary<string, object>());
+                }
+
+                AddAnnotationsToDictionary(deserializerProvider, readContext, instAnnotationDictionary, oDataResource.InstanceAnnotations, annotationKey);
+                shouldSetInstanceAnnotations = true;
+            }
+
+            foreach(ODataProperty property in oDataResource.Properties)
+            {
+                if(property.InstanceAnnotations != null && property.InstanceAnnotations.Count > 0)
+                {
+                    string annotationKey = property.Name;
+
+                    if (!instAnnotationDictionary.ContainsKey(annotationKey))
+                    {
+                        instAnnotationDictionary.Add(annotationKey, new Dictionary<string, object>());
+                    }
+
+                    AddAnnotationsToDictionary(deserializerProvider, readContext, instAnnotationDictionary, property.InstanceAnnotations, annotationKey);
+                    shouldSetInstanceAnnotations = true;
+                }
+            }
+
+            if (shouldSetInstanceAnnotations)
+            {
+                SetProperty(resource, annotationPropertyName, instAnnotationDictionary);
+            }
```

Comment:
Shouldn't the instanceAnnotationDictionary already be set on the resource by the time this method is called?  Prior to calling SetInstanceAnnotations we should either get or create the instanceAnnotationsDictionary on the resource, then we shouldn't have to pass the resource or annotationPropertyName to this method, just the instanceAnnotationsDictionary.